File: sdk/php/src/Internal/Infra/DefaultWsTokenProvider.php
<?php

namespace KuCoin\UniversalSDK\Internal\Infra;

use Exception;
use KuCoin\UniversalSDK\Internal\Interfaces\Transport;
use KuCoin\UniversalSDK\Internal\Interfaces\WsTokenProvider;

/**
 * Default implementation of the WebSocket token provider
 * Handles token retrieval and management for WebSocket connections
 */
class DefaultWsTokenProvider implements WsTokenProvider
{
    private const PATH_PRIVATE = '/api/v1/bullet-private';
    private const PATH_PUBLIC = '/api/v1/bullet-public';

    /** @var Transport $transport */
    private $transport;

    /** @var string */
    private $domainType;

    /** @var bool */
    private $isPrivate;

    public function __construct(Transport $transport, string $domainType, bool $isPrivate)
    {
        $this->transport = $transport;
        $this->domainType = $domainType;
        $this->isPrivate = $isPrivate;
    }

    /**
     * Retrieves WebSocket tokens from the server
     * @return WsToken[]
     * @throws Exception
     */
    public function getToken(): array
    {
        $path = $this->isPrivate ? $this::PATH_PRIVATE : $this::PATH_PUBLIC;

        /** @var TokenResponse $result */
        $result = $this->transport->call(
            $this->domain,
            false,
            'POST',
            $path,
            null,
            TokenResponse::class,
            false
        );

        if (!empty($result->instanceServers) && !empty($result->token)) {
            foreach ($result->instanceServers as $server) {
                $server->token = $result->token;
            }
            return $result->instanceServers;
        }

        return [];
    }

    /**
     * Closes the token provider and its associated transport
     */
    public function close()
    {
        $this->transport->close();
    }
}

File: sdk/java/src/main/java/com/kucoin/universal/sdk/internal/infra/DefaultWsTokenProvider.java
package com.kucoin.universal.sdk.internal.infra;

import com.kucoin.universal.sdk.internal.interfaces.Response;
import com.kucoin.universal.sdk.internal.interfaces.Transport;
import com.kucoin.universal.sdk.internal.interfaces.WsToken;
import com.kucoin.universal.sdk.internal.interfaces.WsTokenProvider;
import com.kucoin.universal.sdk.model.RestResponse;
import java.util.Collections;
import java.util.List;
import java.util.Objects;
import lombok.Data;
import lombok.ToString;

public class DefaultWsTokenProvider implements WsTokenProvider {

  private static final String PATH_PRIVATE = "/api/v1/bullet-private";
  private static final String PATH_PUBLIC = "/api/v1/bullet-public";

  private final Transport transport;
  private final String domainType;
  private final boolean isPrivate;

  public DefaultWsTokenProvider(Transport transport, String domainType, boolean isPrivate) {
    this.transport = Objects.requireNonNull(transport, "transport");
    this.domainType = Objects.requireNonNull(domainType, "domainType");
    this.isPrivate = isPrivate;
  }

  @Override
  public List<WsToken> getToken() {
    String path = isPrivate ? PATH_PRIVATE : PATH_PUBLIC;

    TokenResponse result =
        transport.call(domainType, false, "POST", path, null, TokenResponse.class, false);

    if (result == null
        || result.getToken() == null
        || result.getInstanceServers() == null
        || result.getInstanceServers().isEmpty()) {
      return Collections.emptyList();
    }

    result.getInstanceServers().forEach(s -> s.setToken(result.getToken()));
    return result.getInstanceServers();
  }

  @Override
  public void close() {
    transport.close();
  }

  @Data
  @ToString
  private static class TokenResponse
      implements Response<TokenResponse, RestResponse<TokenResponse>> {
    private String token;
    private List<WsToken> instanceServers;

    @Override
    public void setCommonResponse(RestResponse<TokenResponse> response) {}
  }
}
